docs/proposals/component-registration.md

```diff
@@ -0,0 +1,337 @@
+	Type    ComponentStatusesConditionType `json:"type"`
+	Status  ConditionStatus                `json:"status"`
+	Message string                         `json:"message,omitempty"`
+	Error   string                         `json:"error,omitempty"`
+}
+
+// ComponentStatuses (and ComponentStatusesList) holds the cluster validation info.
+// Deprecated: replaced by ComponentStatus
+type ComponentStatuses struct {
+	TypeMeta   `json:",inline"`
+	ObjectMeta `json:"metadata,omitempty"`
+
+	Conditions []ComponentStatusesCondition `json:"conditions,omitempty"`
+}
+
+// ComponentStatusesList describes a list of component statuses.
+// Deprecated: replaced by ComponentList
+type ComponentStatusesList struct {
+	TypeMeta `json:",inline"`
+	ListMeta `json:"metadata,omitempty"`
+
+	Items []ComponentStatuses `json:"items"`
+}
+```
+
+### Kubectl (CLI)
+
+Add a `kubectl get components` command to list the components and their last known conditions.
+
+### Storage
+
+Store component spec/status in etcd, with an apiserver storage wrapper.
+
+### Status Updates
+
+Add a component-controller that regularly probes registered components and updates their stored last known status.
+
+Add self-registration to the components that uses the new Components API (Create call) when the component starts.
+
+Add self-deregistration to the components that uses the new Components API (Remove call) when the component stops gracefully (e.g. after receiving SIGTERM).
+
+Add self-updating to the components that uses the new Components API (Update call) when the component errors/panics.
+
+Use a TCP probe to check liveness, because it indicates a server is listening on the expected host/path, even if its responses are unhealthy.
+
+Update individual component `/healthz` endpoints to perform as an indicator of readiness, not just liveness. Readiness should include transitively checking the readiness of external dependencies (e.g. etcd or other databases).
+
+
+## History
+
+The `/componentstatuses` endpoint and associated `kubectl get componentstatuses` CLI command were added to facilitate retrieving the status of the cluster.
+
+Limitations of the current implementation:
+
+- Components are hardcoded
+  - Components assumed to be localhost accessible
+    - Not required/enforced by any other method
+    - Incompatible with HA deployments (deploying components on different machines)
+    - Incompatible with non-host docker networking (deploying components in different containers)
+  - Components assumed to use default ports
+    - Incompatible with custom port configurations
+  - Components assumed to have a root `/healthz` endpoint
+    - Incompatible with components behind reverse proxies with path prefixes
+- Etcd is included as a component
+  - Inconsistent with convention that `/healthz` success implies usability (aka, dependencies are also usable, even if not completely healthy)
+  - Indicates a k8s cluster problem even if only 1 of 3 HA Etcd nodes are down
```

Comment:
You're right, the proposal doesn't solve the problem of knowing if a type of HA component is ready as a group, because that requires a policy or business logic specific to a type of component.

However, it's up to the component to know how to check the readiness of its dependencies. If the dependency itself is HA, there's two ways to do this:

1. The component knows about all the nodes in a dependency cluster and hits each node's readiness endpoint to verify that half+1 are ready. 
2. The dependency cluster has a master (which the component keeps track of) that knows cluster health. The component then only needs to hit the dependency's master readiness endpoint to determine if the dependency cluster is ready.

I assume that we will eventually want to support both options for components, but without HA examples and use cases, I'm hesitant to try to define the API.